File: python/base/blockmodel.py
import numpy as np

class BlockModel(object):
    def __init__(self,nx,ny,nz,sx=1.0,sy=1.0,sz=1.0):
        self.nx= nx
        self.ny= ny
        self.nz= nz
        self.nxy= nx*ny
        self.n = self.nxy*nz
        self.sizex = sx
        self.sizey = sy
        self.sizez = sz
        self.volume = sx * sy * sz
        
    def get_blockid(self,i,j,k):
        return k*self.nxy + j*self.nx + i

    def get_volume(self):
        return self.volume

File: cpp/blockcaving/BlockModel.cpp
#include "BlockModel.h"

BlockModel::BlockModel(int _nx,int _ny, int _nz): nx(_nx), ny(_ny), nz(_nz)
{
    this->nxy= nx*ny;
    this->n = nxy*nz;
    this->sizex = 1.0;
    this->sizey = 1.0;
    this->sizez = 1.0;
    this->volume = 1.0;
}

BlockModel::BlockModel()
{
    this->nxy = 0;
    this->n = 0;
    this->sizex = 1.0;
    this->sizey = 1.0;
    this->sizez = 1.0;
    this->volume = 1.0;
}

BlockModel::~BlockModel()
{
    //dtor
}

block_index BlockModel::getBlockId(int i, int j, int k){
    return k*nxy + j*nx + i;
}

double BlockModel::getVolume() {
    return this->volume;
}

void BlockModel::setVolume(float sizex,float sizey,float sizez) {
    this->sizex = sizex;
    this->sizey = sizey;
    this->sizez = sizez;
    this->volume = sizex * sizey * sizez;
}
